LaTeX/Inhalt/04_Inhalt/unit_tests.tex

```diff
@@ -1 +1,25 @@
+
+\subsection{Professional}
+Der Testcode von Morik ist professionell geschrieben. Dies bedeutet, dass er leicht verständlich ist aufgrund von sprechenden Bezeichnern und dem Befolgen der AAA-Normalform. Außerdem werden keine unrelevanten Aspekte der Applikation getestet, sondern wir fokussieren uns, wie bereits bei ''Thorough'' genannt, auf das Testen der wichtigsten Funktionalitäten.
+
+\section{Code Coverage}
+Wie zuvor bereits genannt, ermöglicht die Code Coverage eine Einschätzung wie ''gründlich'' die Tests tatsächlich sind, indem gemessen wird wie viel Code beim Testen durchlaufen wird im Vergleich zu wie viel Code insgesamt in der kompletten Applikation durchlaufen werden kann. Hierbei werden zwei Metriken besonders häufig verwendet. Zum einen gibt es die Line Coverage, die die ausgeführten Codezeilen zählt, während es zum anderen die Branch Coverage gibt. Diese zählt die ausgeführten Verzweigungsmöglichkeiten. Dabei ist letztere Metrik aussagekräftiger für die tatsächliche ''Gründlichkeit'' der Tests, während die Line Coverage allerdings leichter zu messen ist.\\
+Morik erreicht mit den Tests eine Line Coverage von 53\%, da durch das Ausführen der Tests 521 der insgesamt 983 Codezeilen durchlaufen werden. \todo{Zahlen anpassen} Ermittelt werden diese Zahlen durch das Aufrufen des Dashboard Clients von CTest direkt nach dem Ausführen der Tests\footnote{siehe \url{https://github.com/moorts/Morik/blob/main/run_tests.sh\#L11}}. Dadurch hat man zusätzlich zu der Übersicht welche Tests erfolgreich waren und welche nicht direkt noch die Auskunft über die momentane Line Coverage.
+
+\section{Mocks}
+Einige der Klassen von Morik sind abhängig von anderen Klassen, um richtig funktionieren zu können. Um diese Klassen jedoch trotzdem testen zu können, ohne vorher alle Abhängigkeiten konstruieren zu müssen, werden an ausgewählten Stellen Mock-Objekte eingesetzt. Für das Erzeugen der Mock-Objekte wird \textit{gmock} verwendet, das in GoogleTest enthalten ist. So kann in der Klasse \href{https://github.com/moorts/Morik/blob/main/src/tests/application/AbstractDatabaseInterfaceMock.h}{\textit{AbstractDatabaseInterfaceMock}} bspw. das Mocken eines AbstractDatabaseInterface beobachtet werden, sodass in \href{https://github.com/moorts/Morik/blob/main/src/tests/application/repository_tests.cpp}{\textit{repository\_tests}} die EntryRepository getestet werden kann, die die Übergabe eines AbstractDatabaseInterface im Konstruktor erwartet. Ein weiteres Mock-Objekt ist das \href{https://github.com/moorts/Morik/blob/main/src/tests/adapters/database/AbstractSqlDatabaseMock.h}{\textit{AbstractSqlDatabaseMock}}, das eine AbstractSqlDatabase mockt. Dies wird verwendet, um in \href{https://github.com/moorts/Morik/blob/main/src/tests/adapters/database/databaseInterface_tests.cpp}{\textit{databaseInterface\_tests}} den DatabaseInterface-Adapter testen zu können, der wiederum die Übergabe einer AbstractSqlDatabase im Konstruktor erwartet.
```

Comment:
Perhaps explain broadly how mock classes work (like how return values are manually set and such) and how the mock classes are configured in the tests...

This should be brief though, if you include it at all

Reply:
I think we can see the matters explained to us in the lecture as given knowledge. I would therefore assume that the reader knows how mock objects work.